File: php/native/HelloWorldKomencoNative/QuantumCircuit.php
<?php

namespace Automatski\Komenco;

require_once 'Operation.php';

use Automatski\Komenco\Operation;
use InvalidArgumentException;

class QuantumCircuit {
    private $numQubits;
    private $operations;

    public function __construct(int $numQubits) {
        $this->numQubits = $numQubits;
        $this->operations = [];
    }

    public function getNumQubits(): int {
        return $this->numQubits;
    }

    public function getOperations(): array {
        return $this->operations;
    }

    private function addSingle(string $gate, array $params, array $qubits) {
        if (count($qubits) != 1) {
            throw new \InvalidArgumentException("Number of qubits for gate $gate has to be one");
        }
        foreach ($qubits as $qubit) {
            if ($qubit < 0 || $qubit >= $this->numQubits) {
                throw new \InvalidArgumentException("Invalid qubit: $qubit");
            }
        }
        $this->operations[] = new Operation($gate, $qubits, $params);
    }

    private function addDouble(string $gate, array $params, array $qubits) {
        if (count($qubits) != 2) {
            throw new \InvalidArgumentException("Number of qubits for gate $gate has to be two");
        }
        foreach ($qubits as $qubit) {
            if ($qubit < 0 || $qubit >= $this->numQubits) {
                throw new \InvalidArgumentException("Invalid qubit: $qubit");
            }
        }
        $this->operations[] = new Operation($gate, $qubits, $params);
    }

    private function addTriple(string $gate, array $params, array $qubits) {
        if (count($qubits) != 3) {
            throw new \InvalidArgumentException("Number of qubits for gate $gate has to be three");
        }
        foreach ($qubits as $qubit) {
            if ($qubit < 0 || $qubit >= $this->numQubits) {
                throw new \InvalidArgumentException("Invalid qubit: $qubit");
            }
        }
        $this->operations[] = new Operation($gate, $qubits, $params);
    }

    private function addMultiple(string $gate, array $params, array $qubits) {
        $requiredQubits = (int) (log(strlen($gate)) / log(2));
        if (count($qubits) != $requiredQubits) {
            throw new \InvalidArgumentException("Number of qubits for gate $gate has to be $requiredQubits");
        }
        foreach ($qubits as $qubit) {
            if ($qubit < 0 || $qubit >= $this->numQubits) {
                throw new \InvalidArgumentException("Invalid qubit: $qubit");
            }
        }
        $this->operations[] = new Operation($gate, $qubits, $params);
    }

    public function id(int $qubit1) {
        $this->addSingle("id", [], [$qubit1]);
    }

    public function x(int $qubit1) {
        $this->addSingle("x", [], [$qubit1]);
    }

    public function y(int $qubit1) {
        $this->addSingle("y", [], [$qubit1]);
    }

    public function z(int $qubit1) {
        $this->addSingle("z", [], [$qubit1]);
    }

    public function h(int $qubit1) {
        $this->addSingle("h", [], [$qubit1]);
    }

    public function s(int $qubit1) {
        $this->addSingle("s", [], [$qubit1]);
    }

    public function sdg(int $qubit1) {
        $this->addSingle("sdg", [], [$qubit1]);
    }

    public function t(int $qubit1) {
        $this->addSingle("t", [], [$qubit1]);
    }

    public function tdg(int $qubit1) {
        $this->addSingle("tdg", [], [$qubit1]);
    }

    public function rx(float $theta, int $qubit1) {
        $this->addSingle("rx", [$theta], [$qubit1]);
    }

    public function ry(float $theta, int $qubit1) {
        $this->addSingle("ry", [$theta], [$qubit1]);
    }

    public function rz(float $theta, int $qubit1) {
        $this->addSingle("rz", [$theta], [$qubit1]);
    }

    public function u(float $theta, float $phi, float $lam, int $qubit1) {
        $this->addSingle("u", [$theta, $phi, $lam], [$qubit1]);
    }

    public function u1(float $theta, int $qubit1) {
        $this->addSingle("u1", [$theta], [$qubit1]);
    }

    public function u2(float $phi, float $lam, int $qubit1) {
        $this->addSingle("u2", [$phi, $lam], [$qubit1]);
    }

    public function u3(float $theta, float $phi, float $lam, int $qubit1) {
        $this->addSingle("u3", [$theta, $phi, $lam], [$qubit1]);
    }

    public function sx(int $qubit1) {
        $this->addSingle("sx", [], [$qubit1]);
    }

    public function sxdg(int $qubit1) {
        $this->addSingle("sxdg", [], [$qubit1]);
    }

    public function r(float $theta, float $phi, int $qubit1) {
        $this->addSingle("r", [$theta, $phi], [$qubit1]);
    }

    public function cx(int $qubit1, int $qubit2) {
        $this->addDouble("cx", [], [$qubit1, $qubit2]);
    }

    public function cy(int $qubit1, int $qubit2) {
        $this->addDouble("cy", [], [$qubit1, $qubit2]);
    }

    public function cz(int $qubit1, int $qubit2) {
        $this->addDouble("cz", [], [$qubit1, $qubit2]);
    }

    public function ch(int $qubit1, int $qubit2) {
        $this->addDouble("ch", [], [$qubit1, $qubit2]);
    }

    public function ucrx(float $theta, int $qubit1, int $qubit2) {
        $this->addDouble("ucrx", [$theta], [$qubit1, $qubit2]);
    }

    public function ucry(float $theta, int $qubit1, int $qubit2) {
        $this->addDouble("ucry", [$theta], [$qubit1, $qubit2]);
    }

    public function ucrz(float $theta, int $qubit1, int $qubit2) {
        $this->addDouble("ucrz", [$theta], [$qubit1, $qubit2]);
    }

    public function crx(float $theta, int $qubit1, int $qubit2) {
        $this->addDouble("crx", [$theta], [$qubit1, $qubit2]);
    }

    public function cry(float $theta, int $qubit1, int $qubit2) {
        $this->addDouble("cry", [$theta], [$qubit1, $qubit2]);
    }

    public function crz(float $theta, int $qubit1, int $qubit2) {
        $this->addDouble("crz", [$theta], [$qubit1, $qubit2]);
    }

    public function cr(float $theta, float $phi, float $lam, int $qubit1, int $qubit2) {
        $this->addDouble("cr", [$theta, $phi, $lam], [$qubit1, $qubit2]);
    }

    public function cu(float $theta, float $phi, float $lam, float $gamma, int $qubit1, int $qubit2) {
        $this->addDouble("cu1", [$theta, $phi, $lam, $gamma], [$qubit1, $qubit2]);
    }

    public function cu1(float $theta, int $qubit1, int $qubit2) {
        $this->addDouble("cu1", [$theta], [$qubit1, $qubit2]);
    }

    public function cu2(float $phi, float $lam, int $qubit1, int $qubit2) {
        $this->addDouble("cu2", [$phi, $lam], [$qubit1, $qubit2]);
    }

    public function cu3(float $theta, float $phi, float $lam, int $qubit1, int $qubit2) {
        $this->addDouble("cu3", [$theta, $phi, $lam], [$qubit1, $qubit2]);
    }

    public function dcx(int $qubit1, int $qubit2) {
        $this->addDouble("dcx", [], [$qubit1, $qubit2]);
    }

    public function ecr(int $qubit1, int $qubit2) {
        $this->addDouble("ecr", [], [$qubit1, $qubit2]);
    }

    public function iswap(int $qubit1, int $qubit2) {
        $this->addDouble("iswap", [], [$qubit1, $qubit2]);
    }

    public function rxx(float $theta, int $qubit1, int $qubit2) {
        $this->addDouble("rxx", [$theta], [$qubit1, $qubit2]);
    }

    public function ryy(float $theta, int $qubit1, int $qubit2) {
        $this->addDouble("ryy", [$theta], [$qubit1, $qubit2]);
    }

    public function rzz(float $theta, int $qubit1, int $qubit2) {
        $this->addDouble("rzz", [$theta], [$qubit1, $qubit2]);
    }

    public function rzx(float $theta, int $qubit1, int $qubit2) {
        $this->addDouble("rzx", [$theta], [$qubit1, $qubit2]);
    }

    public function cswap(int $qubit1, int $qubit2, int $qubit3) {
        $this->addTriple("cswap", [], [$qubit1, $qubit2, $qubit3]);
    }

    public function ccx(int $qubit1, int $qubit2, int $qubit3) {
        $this->addTriple("ccx", [], [$qubit1, $qubit2, $qubit3]);
    }

    public function ccy(int $qubit1, int $qubit2, int $qubit3) {
        $this->addTriple("ccy", [], [$qubit1, $qubit2, $qubit3]);
    }

    public function ccz(int $qubit1, int $qubit2, int $qubit3) {
        $this->addTriple("ccz", [], [$qubit1, $qubit2, $qubit3]);
    }

    public function c3x(array $qubits) {
        $this->addMultiple("c3x", [], $qubits);
    }

    public function c4x(array $qubits) {
        $this->addMultiple("c4x", [], $qubits);
    }

    public function mcx(array $qubits) {
        $this->addMultiple("mcx", [], $qubits);
    }

    public function mcu1(float $theta, array $qubits) {
        $this->addMultiple("mcu1", [$theta], $qubits);
    }

    public function mcu2(float $phi, float $lam, array $qubits) {
        $this->addMultiple("mcu2", [$phi, $lam], $qubits);
    }

    public function mcu3(float $theta, float $phi, float $lam, array $qubits) {
        $this->addMultiple("mcu3", [$theta, $phi, $lam], $qubits);
    }

    public function mcz(array $qubits) {
        $this->addMultiple("mcz", [], $qubits);
    }

    public function mcp(array $qubits) {
        $this->addMultiple("mcp", [], $qubits);
    }

    public function mcrx(array $qubits) {
        $this->addMultiple("mcrx", [], $qubits);
    }

    public function mcry(array $qubits) {
        $this->addMultiple("mcry", [], $qubits);
    }

    public function mcrz(array $qubits) {
        $this->addMultiple("mcrz", [], $qubits);
    }

    public function qft(array $qubits) {
        $this->addMultiple("qft", [], $qubits);
    }

    public function inverse_qft(array $qubits) {
        $this->addMultiple("iqft", [], $qubits);
    }

    public function swap(int $qubit1, int $qubit2) {
        $this->addDouble("swap", [], [$qubit1, $qubit2]);
    }

    public function csx(int $qubit1, int $qubit2) {
        $this->addDouble("csx", [], [$qubit1, $qubit2]);
    }

    public function p(float $theta, int $qubit1) {
        $this->addSingle("p", [$theta], [$qubit1]);
    }

    public function cp(float $theta, int $qubit1, int $qubit2) {
        $this->addDouble("cp", [$theta], [$qubit1, $qubit2]);
    }

    public function ccp(float $theta, int $qubit1, int $qubit2, int $qubit3) {
        $this->addTriple("ccp", [$theta], [$qubit1, $qubit2, $qubit3]);
    }

    public function sqrt_x(int $qubit1) {
        $this->addSingle("sqrt_x", [], [$qubit1]);
    }

    public function sqrt_y(int $qubit1) {
        $this->addSingle("sqrt_y", [], [$qubit1]);
    }

    public function sqrt_z(int $qubit1) {
        $this->addSingle("sqrt_z", [], [$qubit1]);
    }

    public function gpi(float $phi, int $qubit1) {
        $this->addSingle("gpi", [$phi], [$qubit1]);
    }

    public function gpi2(float $phi, int $qubit1) {
        $this->addSingle("gpi2", [$phi], [$qubit1]);
    }

    public function xp(float $theta, int $qubit1) {
        $this->addSingle("xp", [$theta], [$qubit1]);
    }

    public function yp(float $theta, int $qubit1) {
        $this->addSingle("yp", [$theta], [$qubit1]);
    }

    public function zp(float $theta, int $qubit1) {
        $this->addSingle("zp", [$theta], [$qubit1]);
    }

    public function xxp(float $theta, int $qubit1, int $qubit2) {
        $this->addDouble("xxp", [$theta], [$qubit1, $qubit2]);
    }

    public function yyp(float $theta, int $qubit1, int $qubit2) {
        $this->addDouble("yyp", [$theta], [$qubit1, $qubit2]);
    }

    public function zzp(float $theta, int $qubit1, int $qubit2) {
        $this->addDouble("zzp", [$theta], [$qubit1, $qubit2]);
    }

    public function phased_xp(float $theta, float $phi, int $qubit1) {
        $this->addSingle("phased_xp", [$theta, $phi], [$qubit1]);
    }

    public function phased_yp(float $theta, float $phi, int $qubit1) {
        $this->addSingle("phased_yp", [$theta, $phi], [$qubit1]);
    }

    public function phased_zp(float $theta, float $phi, int $qubit1) {
        $this->addSingle("phased_zp", [$theta, $phi], [$qubit1]);
    }

    public function cnotp(float $theta, int $qubit1, int $qubit2) {
        $this->addDouble("cnotp", [$theta], [$qubit1, $qubit2]);
    }

    public function ccnotp(float $theta, int $qubit1, int $qubit2, int $qubit3) {
        $this->addTriple("ccnotp", [$theta], [$qubit1, $qubit2, $qubit3]);
    }

    public function cyp(float $theta, int $qubit1, int $qubit2) {
        $this->addDouble("cyp", [$theta], [$qubit1, $qubit2]);
    }

    public function ccyp(float $theta, int $qubit1, int $qubit2, int $qubit3) {
        $this->addTriple("ccyp", [$theta], [$qubit1, $qubit2, $qubit3]);
    }

    public function czp(float $theta, int $qubit1, int $qubit2) {
        $this->addDouble("czp", [$theta], [$qubit1, $qubit2]);
    }

    public function cczp(float $theta, int $qubit1, int $qubit2, int $qubit3) {
        $this->addTriple("cczp", [$theta], [$qubit1, $qubit2, $qubit3]);
    }

    public function measure(array $qubits) {
        $cparams = [];
        $cqubits = [];
        $cgate = "measure";
        if (empty($qubits)) {
            throw new InvalidArgumentException("number of qubits to measure cannot be zero");
        } else {
            foreach ($qubits as $qubit) {
                if ($qubit < 0 || $qubit >= $this->numQubits) {
                    throw new InvalidArgumentException("invalid qubit: " . $qubit);
                } else {
                    $cqubits[] = $qubit;
                }
            }
        }
        $this->operations[] = new Operation($cgate, $cqubits, $cparams);
    }

    public function measureAll() {
        $cparams = [];
        $cqubits = [];
        $cgate = "measure";
        for ($qubit = 0; $qubit < $this->numQubits; $qubit++) {
            $cqubits[] = $qubit;
        }
        $this->operations[] = new Operation($cgate, $cqubits, $cparams);
    }

    public function randomCircuit(int $numberOfOperations, array $qubitsToUseForRandomCircuit) {
        $qubitsCount = count($qubitsToUseForRandomCircuit);
        if ($qubitsCount < 3) {
            throw new InvalidArgumentException("The number of qubits needed to generate random circuits is >= 3");
        }

        $gatesAndQubitsMap = [
            // 1 qubit gates
            "x" => 1, "y" => 1, "z" => 1, "rx" => 1, "ry" => 1, "rz" => 1,
            "h" => 1, "s" => 1, "sdg" => 1, "t" => 1, "tdg" => 1, "sx" => 1, "sxdg" => 1,
            // 2 qubit gates
            "cx" => 2, "cy" => 2, "cz" => 2, "crx" => 2, "cry" => 2, "crz" => 2,
            "ch" => 2, "dcx" => 2, "ecr" => 2, "iswap" => 2, "swap" => 2, "csx" => 2,
            // 3 qubit gates
            "cswap" => 3, "ccx" => 3
        ];

        $allPossibleGates = array_keys($gatesAndQubitsMap);
        $allPossibleGatesCount = count($allPossibleGates);

        for ($i = 0; $i < $numberOfOperations; $i++) {
            $index = array_rand($allPossibleGates);
            $cgate = $allPossibleGates[$index];

            $numberOfQubits = $gatesAndQubitsMap[$cgate];

            shuffle($qubitsToUseForRandomCircuit);
            $cqubits = array_slice($qubitsToUseForRandomCircuit, 0, $numberOfQubits);

            $cparams = [];
            if (in_array($cgate, ["rx", "ry", "rz", "crx", "cry", "crz"])) {
                $theta = mt_rand() / mt_getrandmax() * 2 * M_PI;
                $cparams[] = $theta;
            }

            $this->operations[] = new Operation($cgate, $cqubits, $cparams);
        }
    }

}

?>

File: r/native/HelloWorldKomencoNative/QuantumCircuit.r
# Define the QuantumCircuit class
QuantumCircuit <- setRefClass(
  "QuantumCircuit",
  fields = list(
    numQubits = "numeric",
    operations = "list"
  ),
  methods = list(
    initialize = function(numQubits) {
      .self$numQubits <- numQubits
      .self$operations <- list()
    },
    getNumQubits = function() {
      return(.self$numQubits)
    },
    getOperations = function() {
      return(.self$operations)
    },
    addSingle = function(gate, params, qubits) {
      if (length(qubits) != 1) {
        stop(paste("Number of qubits for gate", gate, "has to be one"))
      }
      for (qubit in qubits) {
        if (qubit < 0 || qubit >= .self$numQubits) {
          stop(paste("Invalid qubit:", qubit))
        }
      }
      .self$operations <- c(.self$operations, list(Operation$new(gate, qubits, params)))
    },
    addDouble = function(gate, params, qubits) {
      if (length(qubits) != 2) {
        stop(paste("Number of qubits for gate", gate, "has to be two"))
      }
      for (qubit in qubits) {
        if (qubit < 0 || qubit >= .self$numQubits) {
          stop(paste("Invalid qubit:", qubit))
        }
      }
      .self$operations <- c(.self$operations, list(Operation$new(gate, qubits, params)))
    },
    addTriple = function(gate, params, qubits) {
      if (length(qubits) != 3) {
        stop(paste("Number of qubits for gate", gate, "has to be three"))
      }
      for (qubit in qubits) {
        if (qubit < 0 || qubit >= .self$numQubits) {
          stop(paste("Invalid qubit:", qubit))
        }
      }
      .self$operations <- c(.self$operations, list(Operation$new(gate, qubits, params)))
    },
    addMultiple = function(gate, params, qubits) {
      requiredQubits <- as.integer(log(length(gate)) / log(2))
      if (length(qubits) != requiredQubits) {
        stop(paste("Number of qubits for gate", gate, "has to be", requiredQubits))
      }
      for (qubit in qubits) {
        if (qubit < 0 || qubit >= .self$numQubits) {
          stop(paste("Invalid qubit:", qubit))
        }
      }
      .self$operations <- c(.self$operations, list(Operation$new(gate, qubits, params)))
    },
    id = function(qubit1) {
      .self$addSingle("id", list(), list(qubit1))
    },
    x = function(qubit1) {
      .self$addSingle("x", list(), list(qubit1))
    },
    y = function(qubit1) {
      .self$addSingle("y", list(), list(qubit1))
    },
    z = function(qubit1) {
      .self$addSingle("z", list(), list(qubit1))
    },
    h = function(qubit1) {
      .self$addSingle("h", list(), list(qubit1))
    },
    s = function(qubit1) {
      .self$addSingle("s", list(), list(qubit1))
    },
    sdg = function(qubit1) {
      .self$addSingle("sdg", list(), list(qubit1))
    },
    t = function(qubit1) {
      .self$addSingle("t", list(), list(qubit1))
    },
    tdg = function(qubit1) {
      .self$addSingle("tdg", list(), list(qubit1))
    },
    rx = function(theta, qubit1) {
      .self$addSingle("rx", list(theta), list(qubit1))
    },
    ry = function(theta, qubit1) {
      .self$addSingle("ry", list(theta), list(qubit1))
    },
    rz = function(theta, qubit1) {
      .self$addSingle("rz", list(theta), list(qubit1))
    },
    u = function(theta, phi, lam, qubit1) {
      .self$addSingle("u", list(theta, phi, lam), list(qubit1))
    },
    u1 = function(theta, qubit1) {
      .self$addSingle("u1", list(theta), list(qubit1))
    },
    u2 = function(phi, lam, qubit1) {
      .self$addSingle("u2", list(phi, lam), list(qubit1))
    },
    u3 = function(theta, phi, lam, qubit1) {
      .self$addSingle("u3", list(theta, phi, lam), list(qubit1))
    },
    sx = function(qubit1) {
      .self$addSingle("sx", list(), list(qubit1))
    },
    sxdg = function(qubit1) {
      .self$addSingle("sxdg", list(), list(qubit1))
    },
    r = function(theta, phi, qubit1) {
      .self$addSingle("r", list(theta, phi), list(qubit1))
    },
    cx = function(qubit1, qubit2) {
      .self$addDouble("cx", list(), list(qubit1, qubit2))
    },
    cy = function(qubit1, qubit2) {
      .self$addDouble("cy", list(), list(qubit1, qubit2))
    },
    cz = function(qubit1, qubit2) {
      .self$addDouble("cz", list(), list(qubit1, qubit2))
    },
    ch = function(qubit1, qubit2) {
      .self$addDouble("ch", list(), list(qubit1, qubit2))
    },
    ucrx = function(theta, qubit1, qubit2) {
      .self$addDouble("ucrx", list(theta), list(qubit1, qubit2))
    },
    ucry = function(theta, qubit1, qubit2) {
      .self$addDouble("ucry", list(theta), list(qubit1, qubit2))
    },
    ucrz = function(theta, qubit1, qubit2) {
      .self$addDouble("ucrz", list(theta), list(qubit1, qubit2))
    },
    crx = function(theta, qubit1, qubit2) {
      .self$addDouble("crx", list(theta), list(qubit1, qubit2))
    },
    cry = function(theta, qubit1, qubit2) {
      .self$addDouble("cry", list(theta), list(qubit1, qubit2))
    },
    crz = function(theta, qubit1, qubit2) {
      .self$addDouble("crz", list(theta), list(qubit1, qubit2))
    },
    cr = function(theta, phi, lam, qubit1, qubit2) {
      .self$addDouble("cr", list(theta, phi, lam), list(qubit1, qubit2))
    },
    cu = function(theta, phi, lam, gamma, qubit1, qubit2) {
      .self$addDouble("cu1", list(theta, phi, lam, gamma), list(qubit1, qubit2))
    },
    cu1 = function(theta, qubit1, qubit2) {
      .self$addDouble("cu1", list(theta), list(qubit1, qubit2))
    },
    cu2 = function(phi, lam, qubit1, qubit2) {
      .self$addDouble("cu2", list(phi, lam), list(qubit1, qubit2))
    },
    cu3 = function(theta, phi, lam, qubit1, qubit2) {
      .self$addDouble("cu3", list(theta, phi, lam), list(qubit1, qubit2))
    },
    dcx = function(qubit1, qubit2) {
      .self$addDouble("dcx", list(), list(qubit1, qubit2))
    },
    ecr = function(qubit1, qubit2) {
      .self$addDouble("ecr", list(), list(qubit1, qubit2))
    },
    iswap = function(qubit1, qubit2) {
      .self$addDouble("iswap", list(), list(qubit1, qubit2))
    },
    rxx = function(theta, qubit1, qubit2) {
      .self$addDouble("rxx", list(theta), list(qubit1, qubit2))
    },
    ryy = function(theta, qubit1, qubit2) {
      .self$addDouble("ryy", list(theta), list(qubit1, qubit2))
    },
    rzz = function(theta, qubit1, qubit2) {
      .self$addDouble("rzz", list(theta), list(qubit1, qubit2))
    },
    rzx = function(theta, qubit1, qubit2) {
      .self$addDouble("rzx", list(theta), list(qubit1, qubit2))
    },
    cswap = function(qubit1, qubit2, qubit3) {
      .self$addTriple("cswap", list(), list(qubit1, qubit2, qubit3))
    },
    ccx = function(qubit1, qubit2, qubit3) {
      .self$addTriple("ccx", list(), list(qubit1, qubit2, qubit3))
    },
    ccy = function(qubit1, qubit2, qubit3) {
      .self$addTriple("ccy", list(), list(qubit1, qubit2, qubit3))
    },
    ccz = function(qubit1, qubit2, qubit3) {
      .self$addTriple("ccz", list(), list(qubit1, qubit2, qubit3))
    },
    c3x = function(qubits) {
      .self$addMultiple("c3x", list(), qubits)
    },
    c4x = function(qubits) {
      .self$addMultiple("c4x", list(), qubits)
    },
    mcx = function(qubits) {
      .self$addMultiple("mcx", list(), qubits)
    },
    mcu1 = function(theta, qubits) {
      .self$addMultiple("mcu1", list(theta), qubits)
    },
    mcu2 = function(phi, lam, qubits) {
      .self$addMultiple("mcu2", list(phi, lam), qubits)
    },
    mcu3 = function(theta, phi, lam, qubits) {
      .self$addMultiple("mcu3", list(theta, phi, lam), qubits)
    },
    mcz = function(qubits) {
      .self$addMultiple("mcz", list(), qubits)
    },
    mcp = function(qubits) {
      .self$addMultiple("mcp", list(), qubits)
    },
    mcrx = function(qubits) {
      .self$addMultiple("mcrx", list(), qubits)
    },
    mcry = function(qubits) {
      .self$addMultiple("mcry", list(), qubits)
    },
    mcrz = function(qubits) {
      .self$addMultiple("mcrz", list(), qubits)
    },
    qft = function(qubits) {
      .self$addMultiple("qft", list(), qubits)
    },
    inverse_qft = function(qubits) {
      .self$addMultiple("iqft", list(), qubits)
    },
    swap = function(qubit1, qubit2) {
      .self$addDouble("swap", list(), list(qubit1, qubit2))
    },
    csx = function(qubit1, qubit2) {
      .self$addDouble("csx", list(), list(qubit1, qubit2))
    },
    p = function(theta, qubit1) {
      .self$addSingle("p", list(theta), list(qubit1))
    },
    cp = function(theta, qubit1, qubit2) {
      .self$addDouble("cp", list(theta), list(qubit1, qubit2))
    },
    ccp = function(theta, qubit1, qubit2, qubit3) {
      .self$addTriple("ccp", list(theta), list(qubit1, qubit2, qubit3))
    },
    sqrt_x = function(qubit1) {
      .self$addSingle("sqrt_x", list(), list(qubit1))
    },
    sqrt_y = function(qubit1) {
      .self$addSingle("sqrt_y", list(), list(qubit1))
    },
    sqrt_z = function(qubit1) {
      .self$addSingle("sqrt_z", list(), list(qubit1))
    },
    gpi = function(phi, qubit1) {
      .self$addSingle("gpi", list(phi), list(qubit1))
    },
    gpi2 = function(phi, qubit1) {
      .self$addSingle("gpi2", list(phi), list(qubit1))
    },
    xp = function(theta, qubit1) {
      .self$addSingle("xp", list(theta), list(qubit1))
    },
    yp = function(theta, qubit1) {
      .self$addSingle("yp", list(theta), list(qubit1))
    },
    zp = function(theta, qubit1) {
      .self$addSingle("zp", list(theta), list(qubit1))
    },
    xxp = function(theta, qubit1, qubit2) {
      .self$addDouble("xxp", list(theta), list(qubit1, qubit2))
    },
    yyp = function(theta, qubit1, qubit2) {
      .self$addDouble("yyp", list(theta), list(qubit1, qubit2))
    },
    zzp = function(theta, qubit1, qubit2) {
      .self$addDouble("zzp", list(theta), list(qubit1, qubit2))
    },
    phased_xp = function(theta, phi, qubit1) {
      .self$addSingle("phased_xp", list(theta, phi), list(qubit1))
    },
    phased_yp = function(theta, phi, qubit1) {
      .self$addSingle("phased_yp", list(theta, phi), list(qubit1))
    },
    phased_zp = function(theta, phi, qubit1) {
      .self$addSingle("phased_zp", list(theta, phi), list(qubit1))
    },
    cnotp = function(theta, qubit1, qubit2) {
      .self$addDouble("cnotp", list(theta), list(qubit1, qubit2))
    },
    ccnotp = function(theta, qubit1, qubit2, qubit3) {
      .self$addTriple("ccnotp", list(theta), list(qubit1, qubit2, qubit3))
    },
    cyp = function(theta, qubit1, qubit2) {
      .self$addDouble("cyp", list(theta), list(qubit1, qubit2))
    },
    ccyp = function(theta, qubit1, qubit2, qubit3) {
      .self$addTriple("ccyp", list(theta), list(qubit1, qubit2, qubit3))
    },
    czp = function(theta, qubit1, qubit2) {
      .self$addDouble("czp", list(theta), list(qubit1, qubit2))
    },
    cczp = function(theta, qubit1, qubit2, qubit3) {
      .self$addTriple("cczp", list(theta), list(qubit1, qubit2, qubit3))
    },
    measure = function(qubits) {
      cparams <- list()
      cqubits <- list()
      cgate <- "measure"
      if (length(qubits) == 0) {
        stop("number of qubits to measure cannot be zero")
      } else {
        for (qubit in qubits) {
          if (qubit < 0 || qubit >= .self$numQubits) {
            stop(paste("invalid qubit:", qubit))
          } else {
            cqubits <- c(cqubits, qubit)
          }
        }
      }
      .self$operations <- c(.self$operations, Operation$new(cgate, cqubits, cparams))
    },
    measureAll = function() {
      cparams <- list()
      cqubits <- list()
      cgate <- "measure"
      for (qubit in 0:(.self$numQubits - 1)) {
        cqubits <- c(cqubits, qubit)
      }
      .self$operations <- c(.self$operations, Operation$new(cgate, cqubits, cparams))
    },
    randomCircuit = function(numberOfOperations, qubitsToUseForRandomCircuit) {
      qubitsCount <- length(qubitsToUseForRandomCircuit)
      if (qubitsCount < 3) {
        stop("The number of qubits needed to generate random circuits is >= 3")
      }

      gatesAndQubitsMap <- list(
        # 1 qubit gates
        x = 1, y = 1, z = 1, rx = 1, ry = 1, rz = 1,
        h = 1, s = 1, sdg = 1, t = 1, tdg = 1, sx = 1, sxdg = 1,
        # 2 qubit gates
        cx = 2, cy = 2, cz = 2, crx = 2, cry = 2, crz = 2,
        ch = 2, dcx = 2, ecr = 2, iswap = 2, swap = 2, csx = 2,
        # 3 qubit gates
        cswap = 3, ccx = 3
      )

      allPossibleGates <- names(gatesAndQubitsMap)
      allPossibleGatesCount <- length(allPossibleGates)

      operations <<- list()
      for (i in 1:numberOfOperations) {
        index <- sample(1:allPossibleGatesCount, 1)
        cgate <- allPossibleGates[index]

        numberOfQubits <- gatesAndQubitsMap[[cgate]]

        qubitsToUseForRandomCircuit <- sample(qubitsToUseForRandomCircuit)
        cqubits <- qubitsToUseForRandomCircuit[1:numberOfQubits]

        cparams <- list()
        if (cgate %in% c("rx", "ry", "rz", "crx", "cry", "crz")) {
          theta <- runif(1, 0, 2 * pi)
          cparams <- c(cparams, theta)
        }

        .self$operations <- c(.self$operations, Operation$new(cgate, cqubits, cparams))
      }
    }
  )
)